scripted-diff: Use scoped enumerations (C++11, "enum class")

practicalswift opened the pull request:
Rationale (from Bjarne Stroustrup's ["C++11 FAQ"](http://www.stroustrup.com/C++11FAQ.html#enum)):

> 
> The enum classes ("new enums", "strong enums") address three problems with traditional C++ enumerations:
> 
> * conventional enums implicitly convert to int, causing errors when someone does not want an enumeration to act as an integer.
> * conventional enums export their enumerators to the surrounding scope, causing name clashes.
> * the underlying type of an enum cannot be specified, causing confusion, compatibility problems, and makes forward declaration impossible.
>
> The new enums are "enum class" because they combine aspects of traditional enumerations (names values) with aspects of classes (scoped members and absence of conversions).

luke-jr commented on src/rpc/protocol.h:
```diff
@@ -29,7 +29,7 @@ enum HTTPStatusCode
 
 //! Bitcoin RPC error codes
-enum RPCErrorCode
+enum class RPCErrorCode
```
Why not just make JSONRPCError accept a RPCErrorCode argument?

practicalswift replied:
@luke-jr Yes, that's my plan. I'll push those changes today hopefully :-)

ArielLorenzo-Luaces replied:
@luke-jr there would still be a few ```code == static_cast<int>(RPCErrorCode:: ...)```. But I agree it's not that many.

However changing JSONRPCError is kicking the can down to Pair where the conversion to int has to be done, or down to UniValue. It just seems to me like RPCErrorCode implicitly should be treated like an int since it will almost always be converted into one.

practicalswift replied:
@luke-jr @ArielLorenzo-Luaces 

Signature now changed from:

```diff
-UniValue JSONRPCError(int code, const std::string& message)
+UniValue JSONRPCError(RPCErrorCode code, const std::string& message)
```

ArielLorenzo-Luaces replied:
I still think that ```RPCErrorCode``` and ```HTTPStatusCode``` are intrinsically different than the other enumerators, since they are sent over HTTP. They are also the only two that have explicit values assigned. They should be treated as integers always and they should require no conversion.

I would have suggested to hide conversion in encode/decode functions. Then the conversion would only have to be done once after receiving an http status. But then I see ```response.status >= 400``` in ```bitcoin-cli.cpp```. This suggests that HTTP status codes are not treated like enumerators and should not be compared to enumerators either.

```RPCErrorCode``` and ```HTTPStatusCode``` should be treated more like macros instead.
Strong typing them would case unnecessary complexity due to the casting. But that would only be a problem if they begin to be used much more often, which is speculation on my part. So these changes are fine for now.

practicalswift commented:
Rebased!

practicalswift commented:
Rebased again! :-)

practicalswift commented:
Rebased! :-)

practicalswift commented:
Rebased. The build error seems unrelated. Could somebody trigger a new Travis CI build? :-)

practicalswift commented:
Anyone willing to review? :-)

promag commented:
@practicalswift commit a338ad0 fails to build. See https://github.com/bitcoin/bitcoin/pull/10607#issuecomment-319955888.

What about casting first?

practicalswift commented:
@promag Right now casting is done first?

```
commit f849a1579df933b4ddefc2486e1459831d693b66
    Cast from HTTPStatusCode to int where required
commit a338ad0ae69aadc1d0f4d62cfd2e8dae0317c769
    scripted-diff: Convert HTTPStatusCode to a scoped enum (C++11)

    -BEGIN VERIFY SCRIPT-
    sed -i 's/enum HTTPStatusCode/enum class HTTPStatusCode/g' src/rpc/protocol.h
    sed -i 's/\(HTTP_OK\|HTTP_BAD_REQUEST\|HTTP_UNAUTHORIZED\|HTTP_FORBIDDEN\|HTTP_NOT_FOUND\|HTTP_BAD_METHOD\|HTTP_INTERNAL_SERVER_ERROR\|HTTP_SERVICE_UNAVAILABLE\)/HTTPStatusCode::\1/g' src/bitcoin-cli.cpp src/httprpc.cpp src/httpserver.cpp src/rest.cpp
    sed -i 's/int nStatus/HTTPStatusCode status/g' src/httprpc.cpp
    sed -i 's/nStatus/status/g' src/httprpc.cpp
    -END VERIFY SCRIPT-
commit b4bc819716981f80108f852c93c245a35159d3a8
    Add WriteReply(HTTPStatusCode, ...) in addition to WriteReply(int, ...)
```

Commits should be atomic if possible, but the combination of a scripted-diff and two dependent manual commits seems be tricky to get atomic. Do you have any suggestions on how to fix it? A unified diff would be ideal :-)

MarcoFalke commented:
Would it be possible to reduce the number of commits, as each of them has to be reviewed separately?

practicalswift commented:
@MarcoFalke Sure, what about doing these changes in in three parts?

* 1. `HTTPStatusCode` (more complicated due to dependence on two manual add-on commits)
* 2. `JSONRPCError` (more complicated due to dependence on one manual add-on commit)
* 3. The rest of the enums (purely scripted-diffs)

A total of six commits. Sounds good?

jgarzik commented:
This change appears to add a lot of redundancy:  It makes that which can be deduced by the compiler explicit at each use site (kinda the opposite of the `auto` trend).

practicalswift commented:
@jgarzik Is the argument that strongly typed enums should never be used, or that strongly typed enums are not appropriate in these specific cases (this PR)? :-)

TheBlueMatt commented:
Strong Concept ACK, explicit is always better than implicit. Maybe start with adding this to the style guide and requiring it for new code?

ryanofsky commented:
> @jgarzik Is the argument that strongly typed enums should never be used (generally), or that strongly typed enums are not appropriate in these specific cases (this PR)? :-)

A simple way to get rid of some of the redundancy would be to remove prefixes from the enum values (e.g. `HTTPStatusCode::HTTP_UNAUTHORIZED` -> `HTTPStatusCode::UNAUTHORIZED`, `RPCErrorCode::RPC_PARSE_ERROR` -> RPCErrorCode::PARSE_ERROR`, etc)

practicalswift commented:
@ryanofsky Good idea! Should I do that in this PR or in a follow up PR? :-)

practicalswift commented:
@MarcoFalke Good point! Squashed into three commits!

MarcoFalke commented:
@practicalswift The suggestion to remove the prefix could be done in this pull. I imagine it easily couples with the scripted diff.

practicalswift commented:
@MarcoFalke Great! I'll fix a scripted-diff for that!

practicalswift commented:
I've now added a scripted-diff which removes redundant prefixes from enum values as suggested by @MarcoFalke. Please review :-)

Changes made:
* `BlockSource::BLOCK_SOURCE_FOO` → `BlockSource::FOO`
* `DBErrors::DB_FOO` → `DBErrors::FOO`
* `FlushStateMode::FLUSH_STATE_FOO` → `FlushStateMode::FOO`
* `HTTPStatusCode::HTTP_FOO` → `HTTPStatusCode::FOO`
* `HelpMessageMode::HMM_FOO` → `HelpMessageMode::FOO`
* `RBFTransactionState::RBF_TRANSACTIONSTATE_FOO` → `RBFTransactionState::FOO`
* `RPCErrorCode::RPC_FOO` → `RPCErrorCode::FOO`
* `RetFormat::RF_FOO` → `RetFormat::FOO`
* `SigVersion::SIGVERSION_FOO` → `SigVersion::FOO`
* `ThresholdState::THRESHOLD_FOO` → `ThresholdState::FOO`
* `txnouttype::TX_FOO` → `txnouttype::FOO`
* `WitnessMode::WITNESS_FOO` → `WitnessMode::FOO`

practicalswift commented:
Rebased!

practicalswift commented:
Rebased!

TheBlueMatt commented:
Hmm, I dont believe contrib/devtools/commit-script-check.sh supports multiple scripts in a single commit. Personally I'm fine with them being independant commits and dont really mind a 20-commit PR for something like this, but if you want to keep it squashed, you could just make it all one script and put comments in it.

practicalswift commented:
@TheBlueMatt Ah, thanks for letting me know! I've now put it into one script with comments.

practicalswift commented:
Rebased! Anyone willing to review? Perhaps @laanwj? :-)

TheBlueMatt commented:
Notes on the first script: why the extra "\([^:]\)" in the second sed? It seems to run identically without it?
The s/== FLUSH_STATE_/== FlushStateMode::FLUSH_STATE_/g replace seems redundant.

It would be nice to make it build after the first commit. You could probably just move the two fix commits to before the first scripted-diff.

practicalswift commented:
@TheBlueMatt Thanks for reviewing!

The `([^:])` was left from an earlier attempt att making the scripted-diff idempotent :-)  Now removed.

Removed the redundant `FLUSH_STATE_`-replacement and re-arranged the commits.

Looks good now? :-)

practicalswift commented:
Rebased! :-)

practicalswift commented:
Anyone willing to review? :-)

practicalswift commented:
Rebased!

practicalswift commented:
@laanwj Is there any chance that this one will make it into 0.16? If so, I'll keep it updated/rebased :-)

MarcoFalke commented:
@practicalswift I guess so. It just needs someone to review ;)

sipa commented:
Can you do the cast to int using a C-style cast? Those are preferred for primitive types, and a bit more concise to read.

practicalswift commented:
Now using C-style casts for primitive types as suggested by @sipa :-)

laanwj commented:
@practicalswift I'm sorry but there are too many all-over-the-place "cosmetic" changes like this lately, it kind of tires me, trying to focus my limited time on PRs that add functionality or fix bugs. So no, I don't expect to get around to reviewing this any time soon.

practicalswift commented:
@laanwj I do not see this PR as cosmetic. Traditional C++ enumerations cause real problems (name clashes, implicit conversion to int, inability to forward declare, etc.) and I would argue that C++11 enum classes are strictly better from a secure coding/"minimize surprises" perspective.

FWIW the [C++ Core Guidelines authors seem to agree](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-class) .-)

Hope somebody is willing to review and ACK this scripted-diff PR :-)

practicalswift commented:
Rebased! :-)

practicalswift commented:
Rebased! :-)

practicalswift commented:
Rebased!

practicalswift commented:
Rebased again! :-)

@MarcoFalke @TheBlueMatt – you seemed to be positive to this PR. Would you mind reviewing? :-)

practicalswift commented:
Rebased number #12 performed! :-)

MarcoFalke commented:
Since this doesn't get any review, you might want to leave everything that requires a static_cast for later. And regardless, the goal of `enum class` is to be type safe, whereas for an enum that represents ints, type safety is still preferable, but maybe not as important, since the tradeoff is to blaster the code with static_casts

practicalswift commented:
@MarcoFalke Good idea! Removed changes to `HTTPStatusCode`, `RPCErrorCode` and `Base58Type` which all required `static_cast`:s. The scripted-diffs andthe resulting changes should now be easier to review. Thanks!

practicalswift commented:
Rebase number 13 performed! :-)

practicalswift commented:
Rebase number 14 performed! :-)

MarcoFalke commented:
Suggested squash to one commit and scripted diff:


```
scripted-diff: Convert 15 enums into scoped enums (C++11)

-BEGIN VERIFY SCRIPT-

sed -i 's/enum DBErrors/enum class DBErrors/g' src/wallet/walletdb.h
git grep -l DB_ | xargs sed -i 's/DB_\(LOAD_OK\|CORRUPT\|NONCRITICAL_ERROR\|TOO_NEW\|LOAD_FAIL\|NEED_REWRITE\)/DBErrors::\1/g'
sed -i 's/^    DBErrors::/    /g' src/wallet/walletdb.h

sed -i 's/enum VerifyResult/enum class VerifyResult/g' src/wallet/db.h
sed -i 's/\(VERIFY_OK\|RECOVER_OK\|RECOVER_FAIL\)/VerifyResult::\1/g' src/wallet/db.cpp

sed -i 's/enum ThresholdState/enum class ThresholdState/g' src/versionbits.h
git grep -l THRESHOLD_ | xargs sed -i 's/THRESHOLD_\(DEFINED\|STARTED\|LOCKED_IN\|ACTIVE\|FAILED\)/ThresholdState::\1/g'
sed -i 's/^    ThresholdState::/    /g' src/versionbits.h

sed -i 's/enum SigVersion/enum class SigVersion/g' src/script/interpreter.h
git grep -l SIGVERSION_ | xargs sed -i 's/SIGVERSION_\(BASE\|WITNESS_V0\)/SigVersion::\1/g'
sed -i 's/^    SigVersion::/    /g' src/script/interpreter.h

sed -i 's/enum RetFormat {/enum class RetFormat {/g' src/rest.cpp
sed -i 's/RF_\(UNDEF\|BINARY\|HEX\|JSON\)/RetFormat::\1/g' src/rest.cpp
sed -i 's/^    RetFormat::/    /g' src/rest.cpp

sed -i 's/enum HelpMessageMode {/enum class HelpMessageMode {/g' src/init.h
git grep -l HMM_ | xargs sed -i 's/HMM_BITCOIN/HelpMessageMode::BITCOIN/g' src/init.cpp src/bitcoind.cpp src/qt/utilitydialog.cpp src/init.h
sed -i 's/^    HelpMessageMode::/    /g' src/init.h

sed -i 's/enum FeeEstimateHorizon/enum class FeeEstimateHorizon/g' src/policy/fees.h

sed -i 's/enum RBFTransactionState/enum class RBFTransactionState/g' src/policy/rbf.h
git grep -l RBF_ | xargs sed -i 's/RBF_TRANSACTIONSTATE_\(UNKNOWN\|REPLACEABLE_BIP125\|FINAL\)/RBFTransactionState::\1/g'
sed -i 's/^    RBFTransactionState::/    /g' src/policy/rbf.h

sed -i 's/enum BlockSource {/enum class BlockSource {/g' src/qt/clientmodel.h
git grep -l BLOCK_SOURCE_ | xargs sed -i 's/BLOCK_SOURCE_\(NONE\|REINDEX\|DISK\|NETWORK\)/BlockSource::\1/g'
sed -i 's/^    BlockSource::/    /g' src/qt/clientmodel.h

sed -i 's/enum FlushStateMode {/enum class FlushStateMode {/g' src/validation.cpp
sed -i 's/FLUSH_STATE_\(NONE\|IF_NEEDED\|PERIODIC\|ALWAYS\)/FlushStateMode::\1/g' src/validation.cpp
sed -i 's/^    FlushStateMode::/    /g' src/validation.cpp

sed -i 's/enum WitnessMode {/enum class WitnessMode {/g' src/test/script_tests.cpp
sed -i 's/WITNESS_\(NONE\|PKH\|SH\)/WitnessMode::\1/g' src/test/script_tests.cpp
sed -i 's/^    WitnessMode::/    /g' src/test/script_tests.cpp

-END VERIFY SCRIPT-

practicalswift commented:
@MarcoFalke Good idea! Done!

Please re-review :-)

MarcoFalke commented:
utACK 4ee9a995e13f88668629f8c906ea8d0139f5a6e2

practicalswift commented:
Rebase number 15 performed! :-)

MarcoFalke commented:
re-utACK 0f201feccd23cf195b6570d775fb709b79c8db5e

practicalswift committed:
scripted-diff: Convert 11 enums into scoped enums (C++11)

-BEGIN VERIFY SCRIPT-

sed -i 's/enum DBErrors/enum class DBErrors/g' src/wallet/walletdb.h
git grep -l DB_ | xargs sed -i 's/DB_\(LOAD_OK\|CORRUPT\|NONCRITICAL_ERROR\|TOO_NEW\|LOAD_FAIL\|NEED_REWRITE\)/DBErrors::\1/g'
sed -i 's/^    DBErrors::/    /g' src/wallet/walletdb.h

sed -i 's/enum VerifyResult/enum class VerifyResult/g' src/wallet/db.h
sed -i 's/\(VERIFY_OK\|RECOVER_OK\|RECOVER_FAIL\)/VerifyResult::\1/g' src/wallet/db.cpp

sed -i 's/enum ThresholdState/enum class ThresholdState/g' src/versionbits.h
git grep -l THRESHOLD_ | xargs sed -i 's/THRESHOLD_\(DEFINED\|STARTED\|LOCKED_IN\|ACTIVE\|FAILED\)/ThresholdState::\1/g'
sed -i 's/^    ThresholdState::/    /g' src/versionbits.h

sed -i 's/enum SigVersion/enum class SigVersion/g' src/script/interpreter.h
git grep -l SIGVERSION_ | xargs sed -i 's/SIGVERSION_\(BASE\|WITNESS_V0\)/SigVersion::\1/g'
sed -i 's/^    SigVersion::/    /g' src/script/interpreter.h

sed -i 's/enum RetFormat {/enum class RetFormat {/g' src/rest.cpp
sed -i 's/RF_\(UNDEF\|BINARY\|HEX\|JSON\)/RetFormat::\1/g' src/rest.cpp
sed -i 's/^    RetFormat::/    /g' src/rest.cpp

sed -i 's/enum HelpMessageMode {/enum class HelpMessageMode {/g' src/init.h
git grep -l HMM_ | xargs sed -i 's/HMM_BITCOIN/HelpMessageMode::BITCOIN/g'
sed -i 's/^    HelpMessageMode::/    /g' src/init.h

sed -i 's/enum FeeEstimateHorizon/enum class FeeEstimateHorizon/g' src/policy/fees.h

sed -i 's/enum RBFTransactionState/enum class RBFTransactionState/g' src/policy/rbf.h
git grep -l RBF_ | xargs sed -i 's/RBF_TRANSACTIONSTATE_\(UNKNOWN\|REPLACEABLE_BIP125\|FINAL\)/RBFTransactionState::\1/g'
sed -i 's/^    RBFTransactionState::/    /g' src/policy/rbf.h

sed -i 's/enum BlockSource {/enum class BlockSource {/g' src/qt/clientmodel.h
git grep -l BLOCK_SOURCE_ | xargs sed -i 's/BLOCK_SOURCE_\(NONE\|REINDEX\|DISK\|NETWORK\)/BlockSource::\1/g'
sed -i 's/^    BlockSource::/    /g' src/qt/clientmodel.h

sed -i 's/enum FlushStateMode {/enum class FlushStateMode {/g' src/validation.cpp
sed -i 's/FLUSH_STATE_\(NONE\|IF_NEEDED\|PERIODIC\|ALWAYS\)/FlushStateMode::\1/g' src/validation.cpp
sed -i 's/^    FlushStateMode::/    /g' src/validation.cpp

sed -i 's/enum WitnessMode {/enum class WitnessMode {/g' src/test/script_tests.cpp
sed -i 's/WITNESS_\(NONE\|PKH\|SH\)/WitnessMode::\1/g' src/test/script_tests.cpp
sed -i 's/^    WitnessMode::/    /g' src/test/script_tests.cpp

-END VERIFY SCRIPT-

practicalswift commented:
Rebased! Please re-review :-)

sipa commented:
utACK 1f45e2164a7674f716b425a6658c41ca7c30265b

MarcoFalke commented:
re-utACK 0f201feccd23cf195b6570d775fb709b79c8db5e (confirmed the scripted diff is still the same and the rebase was solved correctly)

practicalswift commented:
Thanks a lot for reviewing @sipa and @MarcoFalke. Excellent!

Anyone else who has time for reviewing in order to bring this PR closer to merge? :-)

practicalswift commented:
@laanwj Would it be possible to get a 0.17.0 milestone for this PR? :-)

laanwj commented:
Probably better to just merge this...
utACK 1f45e21
